https://svn.boost.org/trac/boost/ticket/9879

--- tools/regression/src/compiler_status.cpp
+++ tools/regression/src/compiler_status.cpp
@@ -128,7 +128,7 @@
         string ln(line);
         if (ln.find("Revision: ") != string::npos)
         {
-          for(auto itr = ln.begin()+10; itr != ln.end() && isdigit(*itr); ++itr)
+          for(string::iterator itr = ln.begin()+10; itr != ln.end() && isdigit(*itr); ++itr)
             rev += *itr;
         }
       }
