--- cffi/verifier.py
+++ cffi/verifier.py
@@ -7,10 +7,24 @@
 
     def __init__(self, ffi, preamble, tmpdir=None, modulename=None,
                  ext_package=None, tag='', force_generic_engine=False, **kwds):
+        if os.environ.get("PYTHON_CFFI_DEBUGGING", None) == "1":
+            print("=" * 80)
+            print("Python CFFI module configuration:")
+            if modulename:
+                print("* Module name: %r" % modulename)
+            if tag:
+                print("* Tag: %r" % tag)
+            print("* Source: %r" % preamble)
+            print("* Keyword arguments: %r" % kwds)
+            print("* CDef sources: %r" % ffi._cdefsources)
+            print("=" * 80)
         self.ffi = ffi
         self.preamble = preamble
         if not modulename:
-            flattened_kwds = ffiplatform.flatten(kwds)
+            module_name_affecting_kwds = kwds.copy()
+            module_name_affecting_kwds.pop("include_dirs", None)
+            module_name_affecting_kwds.pop("library_dirs", None)
+            flattened_kwds = ffiplatform.flatten(module_name_affecting_kwds)
         vengine_class = _locate_engine_class(ffi, force_generic_engine)
         self._vengine = vengine_class(self)
         self._vengine.patch_extension_kwds(kwds)
