--- c/minibuffer.h
+++ c/minibuffer.h
@@ -170,7 +170,11 @@
     else if (PySlice_Check(item)) {
         Py_ssize_t start, stop, step, slicelength;
 
+#if PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION < 2
+        if (PySlice_GetIndicesEx((PySliceObject*)item, self->mb_size,
+#else
         if (PySlice_GetIndicesEx(item, self->mb_size,
+#endif
                          &start, &stop, &step, &slicelength) < 0)
             return NULL;
 
@@ -203,7 +207,11 @@
     else if (PySlice_Check(item)) {
         Py_ssize_t start, stop, step, slicelength;
 
+#if PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION < 2
+        if (PySlice_GetIndicesEx((PySliceObject*)item, self->mb_size,
+#else
         if (PySlice_GetIndicesEx(item, self->mb_size,
+#endif
                          &start, &stop, &step, &slicelength) < 0) {
             return -1;
         }
