Revert r1145203.

--- subversion/bindings/swig/perl/native/Makefile.PL.in
+++ subversion/bindings/swig/perl/native/Makefile.PL.in
@@ -72,7 +72,7 @@
 my %config = (
     ABSTRACT => 'Perl bindings for Subversion',
     DEFINE => $cppflags,
-    CCFLAGS => join(' ', $cflags, $Config{ccflags}),
+    CCFLAGS => $cflags,
     INC  => join(' ', $includes, $cppflags,
                  " -I$swig_srcdir/perl/libsvn_swig_perl",
                  " -I$svnlib_srcdir/include",
